src/Libraries/CoreNodeModels/Remember.cs

```diff
@@ -0,0 +1,126 @@
+    [OutPortDescriptions(typeof(Properties.Resources), nameof(Properties.Resources.RememberOuputToolTip))]
+    [IsDesignScriptCompatible]
+    [DynamoServices.RegisterForTrace]
+    public class Remember : NodeModel
+    {
+        private string cache = "";
+
+        [JsonProperty("Cache")]
+        public string Cache
+        {
+            get { return cache; }
+            set
+            {
+                var valueToSet = value == null ? "" : value;
+                if (valueToSet != cache)
+                {
+                    cache = valueToSet;
+                    MarkNodeAsModified();
+                }
+            }
+        }
+
+        [JsonConstructor]
+        private Remember(IEnumerable<PortModel> inPorts, IEnumerable<PortModel> outPorts) : base(inPorts, outPorts)
+        {
+            PropertyChanged += OnPropertyChanged;
+        }
+
+        public Remember()
+        {
+            RegisterAllPorts();
+            PropertyChanged += OnPropertyChanged;
+        }
+
+        private void OnPropertyChanged(object sender, PropertyChangedEventArgs e)
+        {
+            switch (e.PropertyName)
+            {
+                case "State":
+                    if (State == ElementState.Warning)
+                    {
+                        Cache = "";
+                    }
+                    break;
+
+                default:
+                    // Nothing to handle
+                    break;
+            }
+        }
+
+        protected override void OnBuilt()
+        {
+            base.OnBuilt();
+            DataBridge.Instance.RegisterCallback(GUID.ToString(), DataBridgeCallback);
+        }
+
+        public override void Dispose()
+        {
+            PropertyChanged -= OnPropertyChanged;
+            base.Dispose();
+            DataBridge.Instance.UnregisterCallback(GUID.ToString());
+        }
+
+        private static readonly string BuiltinDictionaryTypeName = typeof(DesignScript.Builtin.Dictionary).FullName;
+        private static readonly string BuiltinDictionaryGet = nameof(DesignScript.Builtin.Dictionary.ValueAtKey);
+
+        public override IEnumerable<AssociativeNode> BuildOutputAst(List<AssociativeNode> inputAstNodes)
+        {
+            var resultAst = new List<AssociativeNode>();
+
+            var funtionInputs = new List<AssociativeNode> { inputAstNodes[0], AstFactory.BuildStringNode(Cache) };
+
+            //First build the function call
+            var functionCall = AstFactory.BuildFunctionCall(
+               new Func<object, string, Dictionary<string, object>>(DSCore.Data.Remember), funtionInputs);
+
+            var functionCallIndent = AstFactory.BuildIdentifier(GUID + "_func");
+
+            resultAst.Add(AstFactory.BuildAssignment(functionCallIndent, functionCall));
+
+            //Next add the first key value pair to the output port
+            var getFirstKey = AstFactory.BuildFunctionCall(BuiltinDictionaryTypeName, BuiltinDictionaryGet,
```

Comment:
I think the ASTFactory somewhere has methods for building these calls easily.

Reply:
Not that I am aware of or could find.  The main thing we do here is that we need to pass the data from the inputPort and the cached string into and back out of the node every time.  The dictionary is used so that we can only present to the ports data on the node and keep the cache data internal.  I think we look long and hard at the time we wrote this and didn't have any out of the box options of dictionaries in ASTFactory.

Reply:
look for:
DictionaryExpressionBuilder

Reply:
https://github.com/DynamoDS/Dynamo/blob/master/src/Engine/ProtoCore/Parser/DictionaryBuilder.cs

Reply:
Ahhh.  Thanks.  I did look at that.  Interesting but not exactly what we need.  The dictionary is already set up by the function we call `Remember()` as the return type.  The AST construction we are doing is to look up specific keys later...  More specifically `DictionaryBuilder` embeds the constuctor `DesignScript.Builtin.Dictionary.ByKeysValues` whereas we  need the method `DesignScript.Builtin.Dictionary.ValueAtKey`.   There could be a helper for us but not this one.